File: part2/exceptions/trans_excep.py
# Apache Thrift Transport Exceptions in Python

import pickle
import sys
from thrift import Thrift
from thrift.transport import TTransport

class Trade:
    def __init__(self):
        symbol=""
        price=0.0
        size=0

try:
    trans = TTransport.TFileObjectTransport(open("data","wb"))
    trade = Trade()
    trade.symbol = "F"
    trade.price = 13.10
    trade.size = 2500
    trans.write(pickle.dumps(trade));
    trans.close()

    if len(sys.argv) == 2:						
        raise TTransport.TTransportException(
                  TTransport.TTransportException.NOT_OPEN, "cmd line ex")

    trans = TTransport.TFileObjectTransport(open("data", 		
                                    ("wb" if len(sys.argv) > 2 else "rb")))
    bstr = trans.read(128)
    trade = pickle.loads(bstr)
    print("Trade(%d): %s %d @ %f" % (len(bstr), trade.symbol,
                                     trade.size, trade.price))

except TTransport.TTransportException as tte:			
    print("TTransportException(%d): %s" % (tte.type, tte))
except Thrift.TException as te:					
    print("TException: %s" % te)
except Exception as e:						
    print("Exception: %s %s" % (type(e), e))
except:								
    print("BaseException: %s" % sys.exc_info()[0])

File: part2/exceptions/TransExcep.java
// Apache Thrift Transport Exceptions in Java

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import org.apache.thrift.TException;
import org.apache.thrift.transport.TSimpleFileTransport;
import org.apache.thrift.transport.TTransport;
import org.apache.thrift.transport.TTransportException;

public class TransExcep {

    static private class Trade implements Serializable {
        public String symbol;
        public double price;
        public int size;
    };

    public static void main(String[] args) {
        try {
            TTransport trans = new TSimpleFileTransport("data", false, true);
            Trade trade = new Trade();
            trade.symbol = "F";
            trade.price = 13.10;
            trade.size = 2500;
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            ObjectOutputStream oos = new ObjectOutputStream(baos);
            oos.writeObject(trade);
            trans.write(baos.toByteArray());
			trans.close();

            trans = new TSimpleFileTransport("data",
                                         ((args.length==0) ? true : false), 
                                         true);			
            byte[] buf = new byte[128];
            int iBytesRead = trans.read(buf, 0, buf.length);
            ByteArrayInputStream bais = new ByteArrayInputStream(buf);
            ObjectInputStream ois = new ObjectInputStream(bais);
            trade = (Trade) ois.readObject();
            System.out.println("Trade(" + iBytesRead + "): " + trade.symbol + " " + 
                               trade.size + " @ " + trade.price);
        } catch (TTransportException tte) {				
            System.out.println("TTransportException(" + tte.getType() + 
                               "): " + tte);
        } catch (TException te) {						
            System.out.println("TException: " + te);
        } catch (Exception e) {						
            System.out.println("Exception: " + e);
        } catch (Throwable t) {						
            System.out.println("Throwable: " + t);			
        }
    }
}

